### ISSUE TITLE
BROKER_CONNECTION_TIMEOUT not working.

### ISSUE THREAD

**Issue by Raghavsalotra** [celery/celery#4627]
## Checklist

- [ ] I am using celery version 4.1.0
- [ ] I have verified that the issue exists against the `master` branch of Celery.

## Steps to reproduce
Make rabbitmq borker unreachable , while using **Task.delay()** this call should for seconds mentioned in BROKER_CONNECTION_TIMEOUT but it is getting stuck for indefinite seconds.
## Expected behavior

.delay() call should give timeout exception if it is not able to connect to broker.

## Actual behavior
It is stuck in waiting for indefinite time or it could be the default seconds mentioned in kombu which celery is using under the hood to connect to Rabbitmq broker.

**Comment by clokep** [comment]
I've never been able to get a real answer on whether this is the expected behavior or not (see https://github.com/celery/celery/issues/4328#issuecomment-367801326), but I did document that this happens in #4606.

We were able to fix this by monkey patching the kombu `Connection` with the changes from https://github.com/celery/kombu/commit/0f4da8d450d1a44558382c1858f025faea1b3fde and then setting `BROKER_TRANSPORT_OPTIONS` to have fewer retries (although the number of retries you want is fully dependent on your application!)

**Comment by frankwiles** [comment]
I'm seeing this as well.

**Comment by drmclean** [comment]
I have also had this exact issue with celery and RabbitMQ, namely:

`.apply_async()` hangs forver when the broker (RabbitMQ) is down.

The indefinite hang is due to the `retry_over_time()` function in the `Kombu` library. 
(`Kombu.utils.functional.py:retry_over_time()`)

`retry_over_time()` takes a number of arguments:

```
def retry_over_time(fun, catch, args=[], kwargs={}, errback=None,
                    max_retries=None, interval_start=2, interval_step=2,
                    interval_max=30, callback=None):
```

`max_retries` defaults to `None` which means infinite retries (hence infinite hang)

These arguments are set using `broker_transport_options` in: 

`Kombu.connection.py:Connection.default_channel`

```
    @property
    def default_channel(self):
        """Default channel.

        Created upon access and closed when the connection is closed.

        Note:
            Can be used for automatic channel handling when you only need one
            channel, and also it is the channel implicitly used if
            a connection is passed instead of a channel, to functions that
            require a channel.
        """
        conn_opts = {}
        transport_opts = self.transport_options
        if transport_opts:
            if 'max_retries' in transport_opts:
                conn_opts['max_retries'] = transport_opts['max_retries']
            if 'interval_start' in transport_opts:
                conn_opts['interval_start'] = transport_opts['interval_start']
            if 'interval_step' in transport_opts:
                conn_opts['interval_step'] = transport_opts['interval_step']
            if 'interval_max' in transport_opts:
                conn_opts['interval_max'] = transport_opts['interval_max']

        # make sure we're still connected, and if not refresh.
        self.ensure_connection(**conn_opts)
        if self._default_channel is None:
            self._default_channel = self.channel()
        return self._default_channel
```

You can fix the indefinite hang by setting `broker_transport_options` in the celery config.

```
BROKER_TRANSPORT_OPTIONS = {"max_retries": 3, "interval_start": 0, "interval_step": 0.2, "interval_max": 0.5}
```
In this example the broker will retry 3 times, starting immediately, waiting 0.2 seconds more each time and never waiting more than 0.5 seconds so it will retry at:

**+0 seconds**
_((internal = 0+0.2 = 0.2 seconds))_
**+0.2 seconds**
_((internal = 0.2+0.2 = 0.4 seconds))_
**+0.6 seconds**   

I think this is the correct short term solution for the problem unless anyone knows of a better fix?


```
------------------------------------------------------------------------------------
```
**tl;dr**
Put this in your celery config to fix the issue immediately:

```
BROKER_TRANSPORT_OPTIONS = {"max_retries": 3, "interval_start": 0, "interval_step": 0.2, "interval_max": 0.5}
```

**Comment by auvipy** [comment]
may be we could close this then? should the solution needed to be on documentation?

**Comment by drmclean** [comment]
Yes, but also I think the documentation is actually incorrect (or at least very misleading).

looking here: http://docs.celeryproject.org/en/latest/userguide/configuration.html#std:setting-task_publish_retry

The docs suggest that you get set broker-transport options using the `task_publish_retry` and `task_publish_retry_policy`.

The linked page: (http://docs.celeryproject.org/en/latest/userguide/calling.html#calling-retry) even explicity mentions the problems that could occur and the ramifications: 

```
For example, the default policy correlates to:

add.apply_async((2, 2), retry=True, retry_policy={
    'max_retries': 3,
    'interval_start': 0,
    'interval_step': 0.2,
    'interval_max': 0.2,
})
the maximum time spent retrying will be 0.4 seconds. It’s set relatively short by default because a connection failure could lead to a retry pile effect if the broker connection is down – For example, many web server processes waiting to retry, blocking other incoming requests.
```

But those settings don't get applied and you have to use `broker_transport_options` instead - unless I misunderstood the intention of those settings?

Perhaps altering the docs to be more explicit about which settings affect which connections and then even better an overview page with some friendly documentation that covers the difference between:

broker_transport_options

broker_connection_timeout
broker_connection_max_retries

publish_task_retry
publish_task_retry_policy

etc... 
and when to use each? 

What do you think?

**Comment by clokep** [comment]
I tried to improve these a bit in #4606, but I hadn't come across the `retry_policy` before. That still seems like not the best fix though since that is meant to be used at task-call time? I guess it depends on whether you want all your tasks to use the same retry policy or not.

**Comment by drmclean** [comment]
`retry_policy` isn't a fix though. It doesn't get applied even though the docs suggest that it does. 

I think the docs ought to be updated to explicit state that `publish_task_retry` and `publish_task_retry_policy` don't affect how celery will attempt to retry (or not) the publishing of new tasks using `apply_async()`

**Comment by rsyring** [comment]
I'm running into the problem of infinite connection retries when first starting Celery:

Using a bad port number deliberately for demonstration purposes: `amqp://guest@localhost:5679//`
```
$ strace -f -e trace=network pytest -x
strace: Process 11274 attached
...
socket(AF_INET, SOCK_STREAM|SOCK_CLOEXEC, IPPROTO_TCP) = 11
connect(11, {sa_family=AF_INET, sin_port=htons(5679), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
getsockopt(11, SOL_SOCKET, SO_ERROR, [ECONNREFUSED], [4]) = 0
socket(AF_INET, SOCK_STREAM|SOCK_CLOEXEC, IPPROTO_TCP) = 11
connect(11, {sa_family=AF_INET, sin_port=htons(5679), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
getsockopt(11, SOL_SOCKET, SO_ERROR, [ECONNREFUSED], [4]) = 0
socket(AF_INET, SOCK_STREAM|SOCK_CLOEXEC, IPPROTO_TCP) = 11
connect(11, {sa_family=AF_INET, sin_port=htons(5679), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
getsockopt(11, SOL_SOCKET, SO_ERROR, [ECONNREFUSED], [4]) = 0
socket(AF_INET, SOCK_STREAM|SOCK_CLOEXEC, IPPROTO_TCP) = 11
connect(11, {sa_family=AF_INET, sin_port=htons(5679), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
getsockopt(11, SOL_SOCKET, SO_ERROR, [ECONNREFUSED], [4]) = 0
```
The key to me was seeing "sin_port=htons(5679)" repeated over and over again, indicating a repeated attempt to connect to the broker.

I have the config set:

```python

        'broker_transport_options': {'max_retries': 3}
```

But it does not seem to apply.  The connections retry for at least 10 minutes (that's the longest I wanted to wait).

**Comment by Jamim** [comment]
Hello @auvipy,
Just wanted to remind that a milestone is missing for this bug currently.

**Comment by diepmv** [comment]
same issue for celery 4.3.0, kombu 4.6.3

**Comment by pandafy** [comment]
I spent 5 hours debugging my app due to this issue. I want to fix this issue for good. How can I get started with this? Can anyone guide me please?

**Comment by rsyring** [comment]
When I did my research last year, I came across https://github.com/celery/kombu/issues/829 which I thought could be the underlying cause.  However, that issue was closed by  @auvipy, maybe erroneously (?) since people are still running into problems on this issue.

**Comment by pandafy** [comment]
I ran into this today with `celery==4.4.x`. I updated celery after seeing previously closed issues hoping that it might have been fixed in the newer release. I later came across this issue(thanks to a friend on PySlackers Django channel).

The workaround which @drmclean suggested works smoothly but it comes with a cost. 

Reducing number of retries for everything does not help me much. This is my use case which is getting hurt due to this issue. I have a celery task which gets triggered on post_migrate signal. If the broker is unreachable, celery tried 100 time before giving up increasing retry interval each time. This makes the user think the migration command has stalled.

I just want to reduce the number of max_tries to be reduced only for this celery task which gets triggered on post_migrate signal. This is not possible with the workaround suggested by @drmclean. Which brings me to solving this issue which I really want to do. :wink: 

Thanks for the heads up @rsyring. I will start looking from https://github.com/celery/kombu/issues/829 again.

**Comment by MT-Cash** [comment]
@drmclean You suggest to add the following config to fix the issue immediately:

```
BROKER_TRANSPORT_OPTIONS = {"max_retries": 3, "interval_start": 0, "interval_step": 0.2, "interval_max": 0.5}
```

I've seen that `max_retries=3` is the default ([source](https://docs.celeryproject.org/en/stable/reference/celery.app.task.html#celery.app.task.Task.max_retries)), so setting it should not change anything. Do I get something wrong?

What is the default of the other values?

**Comment by MT-Cash** [comment]
I suspect that I might be running into this issue as I see some tasks disappearing. However, I have Celery 5.1.2 and already have those settings:

```
{
    "broker_transport_options": {
        "max_retries": 1,
        "interval_start": 2.0,
        "confirm_publish": True,
    },
    "task_soft_time_limit": 120,
    "task_time_limit": 130,
}
```

Could it still be this issue?

**Comment by MT-Cash** [comment]
Looking at the code, I don't find any place where `interval_max` is actually used. It seems that `wait_for` has a bug.

**Comment by varadganesan** [comment]
I'm using

Celery 4.2.2
Kombu 4.3.0
Redis backend.

I was able to reproduce this issue by:

Logging into my redis console
Executed this command CLIENT PAUSE 120000 WRITE (this commands prevents the writing of anything into redis)
I tried adding BROKER_TRANSPORT_OPTIONS = {
'max_retries': 4,
'interval_start': 0,
'interval_step': 0.5,
'interval_max': 3,
}

I'm still getting the same error, i.e, the worker hangs indefinitely instead of throwing an exception. The StackTrace is as follows:

  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 179, in delay
    return self.apply_async(partial_args, partial_kwargs)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 1232, in apply_async
    return self.run(tasks, body, args, task_id=task_id, **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 1277, in run
    header_result = header(*partial_args, task_id=group_id, **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 953, in __call__
    return self.apply_async(partial_args, **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 978, in apply_async
    args=args, kwargs=kwargs, **options))
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 1054, in _apply_tasks
    **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 1232, in apply_async
    return self.run(tasks, body, args, task_id=task_id, **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 1277, in run
    header_result = header(*partial_args, task_id=group_id, **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 953, in __call__
    return self.apply_async(partial_args, **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 978, in apply_async
    args=args, kwargs=kwargs, **options))
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 1054, in _apply_tasks
    **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/canvas.py", line 218, in apply_async
    return _apply(args, kwargs, **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/app/task.py", line 535, in apply_async
    **options
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/app/base.py", line 745, in send_task
    amqp.send_task_message(P, name, message, **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/celery/app/amqp.py", line 552, in send_task_message
    **properties
  File "/opt/zopim-env/local/lib/python2.7/site-packages/kombu/messaging.py", line 181, in publish
    exchange_name, declare,
  File "/opt/zopim-env/local/lib/python2.7/site-packages/kombu/connection.py", line 510, in _ensured
    return fun(*args, **kwargs)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/kombu/messaging.py", line 203, in _publish
    mandatory=mandatory, immediate=immediate,
  File "/opt/zopim-env/local/lib/python2.7/site-packages/kombu/transport/virtual/base.py", line 608, in basic_publish
    return self._put(routing_key, message, **kwargs)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/kombu/transport/redis.py", line 778, in _put
    client.lpush(self._q_for_pri(queue, pri), dumps(message))
  File "/opt/zopim-env/local/lib/python2.7/site-packages/redis/client.py", line 1333, in lpush
    return self.execute_command('LPUSH', name, *values)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/redis/client.py", line 668, in execute_command
    return self.parse_response(connection, command_name, **options)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/redis/client.py", line 680, in parse_response
    response = connection.read_response()
  File "/opt/zopim-env/local/lib/python2.7/site-packages/redis/connection.py", line 624, in read_response
    response = self._parser.read_response()
  File "/opt/zopim-env/local/lib/python2.7/site-packages/redis/connection.py", line 390, in read_response
    bufflen = recv_into(self._sock, self._buffer)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/redis/_compat.py", line 73, in recv_into
    return _retryable_call(sock, sock.recv_into, *args, **kwargs)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/redis/_compat.py", line 58, in _retryable_call
    return func(*args, **kwargs)
  File "/opt/zopim-env/local/lib/python2.7/site-packages/billiard/pool.py", line 227, in soft_timeout_sighandler
    raise SoftTimeLimitExceeded()
SoftTimeLimitExceeded: SoftTimeLimitExceeded()
The celery worker hangs until a SoftTimeLimitExceeded is thrown. Ideally, I need to know what are the right config params to be added to ensure that an OperationaError is thrown instead of waiting indefinitely until SoftTimeLimitExceeded gets thrown.

**Comment by GaryRevell** [comment]
Hi,

I'm using podman containers and have rabbitMQ running in one and 3 celery workers running in another one. My app is running in another container and is hanging when I execute the apply_async(...) call. 

If I have the celery workers running in the same container as the app then the apply_async(...) calls work.

So running the workers in a separate container seems to be the problem, however I also never get a timeout despite setting up the following values in the celeryconfig.py file:
```
"""
    Celery config file to be used to set up timeout & retry values
"""

BROKER_CONNECTION_TIMEOUT = 2
BROKER_CONNECTION_RETRY = 1
BROKER_CONNECTION_MAX_RETRIES = 1

CELERY_BROKER_CONNECTION_TIMEOUT = 2
CELERY_BROKER_CONNECTION_MAX_RETRIES = 1
CELERY_BROKER_CONNECTION_RETRY = 1

BROKER_TRANSPORT_OPTIONS = {"max_retries": 3, "interval_start": 0, "interval_step": 0.2, "interval_max": 0.5}

```

Any ideas or work arounds please?

Regards

Gary

**Comment by timbo-desosa-overhaul** [comment]
We're using a kafka consumer to `delay` our tasks, and when celery hangs, it takes so long that the consumer gets removed from the consumer group, but also the pod doesn't crash and restart since it's just retrying forever. Would love if we could get some decent timeout options working for this seven year old issue 🤣 

Also someone posted a workaround for this which is to wrap the task delay or apply in a manual timeout function, which is hacky as hell but probably what we'll try next: https://stackoverflow.com/a/71372783
